be/src/formats/orc/apache-orc/c++/include/orc/Common.hh

```diff
@@ -216,40 +217,34 @@ inline bool compare(T val1, T val2) {
 // Specialization for Decimal
 template <>
 inline bool compare(Decimal val1, Decimal val2) {
-    // compare integral parts
-    Int128 integral1 = scaleDownInt128ByPowerOfTen(val1.value, val1.scale);
-    Int128 integral2 = scaleDownInt128ByPowerOfTen(val2.value, val2.scale);
-
-    if (integral1 < integral2) {
-        return true;
-    } else if (integral1 > integral2) {
-        return false;
+    if (val1.scale == val2.scale) {
+        return val1.value < val2.value;
     }
 
-    // integral parts are equal, continue comparing fractional parts
-    // unnecessary to check overflow here because the scaled number will not
-    // exceed original ones
-    bool overflow = false, positive = val1.value >= 0;
-    val1.value -= scaleUpInt128ByPowerOfTen(integral1, val1.scale, overflow);
-    val2.value -= scaleUpInt128ByPowerOfTen(integral2, val2.scale, overflow);
-
-    int32_t diff = val1.scale - val2.scale;
-    if (diff > 0) {
-        val2.value = scaleUpInt128ByPowerOfTen(val2.value, diff, overflow);
+    int128_t value1, value2, res;
+    int32_t delta;
+    bool overflow;
+    if (val1.scale > val2.scale) {
+        value1 = (static_cast<int128_t>(val1.value.getHighBits()) << 64) + val1.value.getLowBits();
+        value2 = (static_cast<int128_t>(val2.value.getHighBits()) << 64) + val2.value.getLowBits();
+        delta = val1.scale - val2.scale;
+        overflow = starrocks::mul_overflow(value2, starrocks::exp10_int128(delta), &res);
         if (overflow) {
-            return positive ? true : false;
+            return value2 >= 0;
+        } else {
+            return value1 < res;
         }
     } else {
-        val1.value = scaleUpInt128ByPowerOfTen(val1.value, -diff, overflow);
+        value1 = (static_cast<int128_t>(val1.value.getHighBits()) << 64) + val1.value.getLowBits();
```

Comment:
L238-L241 is duplicate with L227-L236, we can put this duplicate into a inline function 

static inline compare0(a, b) {
...
}

if (val1.scale > val2.scale) {
    return compare0(val1, val2);
} else {
   return !compare0(val2, val1);
}

Reply:
this is a tiny issue using this way. The function returns true iff `val1 < val2`, so we cannot directly use negation here.

Reply:
done